egs/wsj/s5/steps/dur_model/train.sh

```diff
@@ -0,0 +1,183 @@
+                  '{ for(i=1; i<=NF; i++){ \
+                       n=split($(i), a, " ");\
+                       duration=a[n];\
+                       counts[duration]++;\
+                       if(duration>max_duration) max_duration=duration; \
+                   }} END { for(d=3; d<=max_duration; d++) \
+                              if(counts[d]<min_count){\
+                                print d-1; exit 0;\
+                              }\
+                            print max_duration;\
+                            exit 0;}'` || exit 1;
+
+  fi
+  echo "$0: Max duration: $max_duration"
+
+  $cmd $dir/log/durmod_init.log \
+       durmodel-init --max-duration=$max_duration \
+                     --left-context=$left_context \
+                     --right-context=$right_context \
+                     $phones_dir/roots.int $phones_dir/extra_questions.int \
+                     $durmodel || exit 1;
+
+fi
+
+if [ $stage -le -1 ]; then
+  echo "$0: Preparing examples..."
+  steps/dur_model/get_examples.sh --cmd "$cmd" --shuffle-buffer-size $shuffle_buffer_size $egs_opts $alidir $dir || exit 1;
+fi
+
+
+[ ! -f $durmodel ] && echo "$0: Duration model file not found (have you completed stage -2?): $durmodel" && exit 1;
+[ ! -f $dir/train.egs ] && echo "$0: Train examples file not found (have you completed stage -1?): $dir/train.egs" && exit 1;
+[ ! -f $dir/val.egs ] && echo "$0: Validation examples file not found (have you completed stage -1?): $dir/val.egs" && exit 1;
+
+if $use_gpu; then
+  train_queue_opt="--gpu 1"
+  if ! cuda-compiled; then
+    echo "$0: WARNING: you are running with one thread but you have not compiled"
+    echo "   for CUDA.  You may be running a setup optimized for GPUs.  If you have"
+    echo "   GPUs and have nvcc installed, go to src/ and do ./configure; make"
+    exit 1
+  fi
+else
+  parallel_train_opts="--use-gpu=no"
+fi
+
+if [ $stage -le 0 ]; then  # do the nnet-related initializations only if stage<=0
+  echo "$0: Initializing nnet3 model..."
+  stage=0
+  if [[ ! -z $nnet_config ]]; then
+    echo "$0: Using provided config file for nnet: $nnet_config"
+  else
+    nnet_config=$dir/nnet.conf
+    feat_dim=$(durmodel-info $durmodel | grep feature-dim | awk '{ print $2 }') || exit 1;
+    output_dim=$(durmodel-info $durmodel | grep max-duration | awk '{ print $2 }') || exit 1;
+    steps/dur_model/make_nnet_config.sh $nnet_opts $feat_dim $output_dim >$nnet_config
+    echo "$0: Wrote nnet config to "$nnet_config
+  fi
+  $cmd $dir/log/nnet_init.log \
+       nnet3-init $nnet_config $dir/nnet.raw || exit 1;
+  $cmd $dir/log/nnet_durmodel_init.log \
+       nnet3-durmodel-init $durmodel $dir/nnet.raw $dir/0.mdl || exit 1;
+else
+  [ ! -f $dir/$stage.mdl ] && echo "$0: Nnet-duration model file not found (you provided --stage=$stage): $dir/$stage.mdl" && exit 1;
+fi
+
+num_examples=`cat $dir/num_examples` || num_examples=1
+num_epochs_per_iter=$[40000000/$num_examples]
```

Comment:
Normally in nnet3 training we term an iteration as one training job which processes all the examples in the ark file. An epoch is one pass over all the ark files. So typically there are multiple iterations per epoch. How are iteration and epoch used in this file ?